{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy and paste latest sbhandler.py code\n",
    "\n",
    "# from sbhandler import *\n",
    "from math import pow    \n",
    "# from fastai.vision.all import *\n",
    "\n",
    "# Needed imports\n",
    "import numpy as np\n",
    "import h5py\n",
    "\n",
    "STOP_CODON = np.array([84, 65, 65, 84, 65, 71]) # T A A T A G\n",
    "\n",
    "# Source code for functions from seqblock_handler.py\n",
    "def load_matlab_file(mat_file, variable_name): # Note the double return statement\n",
    "    \"\"\"_summary_\n",
    "    //UNIQUE: Only one variable for this .mat file\n",
    "    \n",
    "    Args:\n",
    "        mat_file (string): Name of .mat file to intake (path must be given if not in current dir)\n",
    "        variable_name (string): which variable from the mat file. Only one here because of the .mat file structure.\n",
    "\n",
    "    Returns:\n",
    "        h5py_object, data: Overall h5py_object and the specific variable as an nparray\n",
    "    \"\"\"\n",
    "    h5py_object = h5py.File(mat_file, 'r')\n",
    "    data = h5py_object.get(variable_name)\n",
    "    data = np.array(data)\n",
    "    return h5py_object, data\n",
    "class DNA_SeqBlocks():\n",
    "    \"\"\"Mat dataset \n",
    "    Loads the mat file with DNA sequences for easy iteration through all stored DNA sequence lists of varying coverage\n",
    "    \"\"\"\n",
    "    def __init__(self, h5py_object, data):\n",
    "        \"\"\"Initialize DNA_SeqBlocks() object\n",
    "        Args:\n",
    "            h5py_object (h5py_object): first return variable of load_matlab_file\n",
    "            data (np.array): second return variable of load_matlab_file\n",
    "        \"\"\"\n",
    "        self.h5py_object = h5py_object\n",
    "        self.data = np.transpose(data)\n",
    "        self.size = len(data) \n",
    "        \n",
    "    def get_seqblock(self, number, true_len = 0, df=False): # checked\n",
    "        if true_len != 0:\n",
    "            x = self.h5py_object[self.data[0, number-1]][0:true_len,:]\n",
    "            if df == True:\n",
    "                return x\n",
    "            return np.transpose(x) # no need to tranpose if its for pandas\n",
    "            \n",
    "        x = self.h5py_object[self.data[0, number-1]][:, :] # //UNIQUE: 0th column because .mat file, align3 variable only has one row\n",
    "        if df == True:\n",
    "            return x\n",
    "        return np.transpose(x)\n",
    "    \n",
    "    def get_target(self, number):\n",
    "        x = self.h5py_object[self.data[0, number-1]][:, 0]\n",
    "        return np.transpose(x)\n",
    "    \n",
    "    def get_coverage_count(self, number):\n",
    "        \"\"\"Coverage count\n",
    "\n",
    "        Args:\n",
    "            number (int): which sequence? matlab idx\n",
    "\n",
    "        Returns:\n",
    "            int: Number of reads\n",
    "        \"\"\"\n",
    "        x = self.h5py_object[self.data[0, number-1]][0,:]\n",
    "        coverage_count = int((len(x)-4)/2)\n",
    "        return coverage_count\n",
    "    \n",
    "    def get_barcode(self, number): # //TODO: Ask Yiyang about a mutated stop codon.\n",
    "        \"\"\"Barcode finder\n",
    "        \n",
    "        Changes that need to be made: use target sequence to find stop codon (reads may have mutated stop codon like 22776). Use read that doesn't terminate in indels to find the barcode.\n",
    "\n",
    "        Args:\n",
    "            number (int): which sequence? matlab indexing\n",
    "\n",
    "        Returns:\n",
    "            barcode, true_len: returns barcode uint16 ndarray, last matlab idx of stop codon\n",
    "        \"\"\"\n",
    "        x = self.h5py_object[self.data[0, number-1]][:, 0]\n",
    "        barcode = \"\"\n",
    "        true_len = 0\n",
    "        total = x.size\n",
    "        \n",
    "        idx = 0\n",
    "        search_stop=True\n",
    "        \n",
    "        # This block is concise but had 2x the runtime\n",
    "        # while idx < total and search_stop: # To identify the stop codon\n",
    "        #     if (np.all(x[idx:idx+6] == STOP_CODON)):  # Translates to 'TAATAG'\n",
    "        #         true_len = idx+6\n",
    "        #         search_stop = False\n",
    "        #     idx += 1\n",
    "        \n",
    "        while idx < total and search_stop:\n",
    "            if (x[idx] == 84):\n",
    "                try:\n",
    "                    if (x[idx+1] == 65):\n",
    "                        if x[idx+2] == 65:\n",
    "                            if x[idx+3] == 84:\n",
    "                                if x[idx+4] == 65:\n",
    "                                    if x[idx+5] == 71:\n",
    "                                        true_len = idx+6\n",
    "                                        search_stop = False\n",
    "                except: \n",
    "                    barcode = None\n",
    "                    true_len = 0\n",
    "                    return barcode, true_len\n",
    "            idx += 1\n",
    "        barcode = x[true_len:total]\n",
    "            \n",
    "        return barcode, true_len\n",
    "    \n",
    "    def get_nuc_qscores(self, number, nuc_pos, coverage_count=0): # checked\n",
    "        if coverage_count == 0:\n",
    "            cov = self.get_coverage_count(number)\n",
    "            x = self.h5py_object[self.data[0, number-1]][nuc_pos-1, (1+cov):(1+2*cov)]\n",
    "            return np.transpose(x)\n",
    "        else:\n",
    "            x = self.h5py_object[self.data[0, number-1]][nuc_pos-1, (1+coverage_count):(1+2*coverage_count)]\n",
    "            return np.transpose(x)\n",
    "    \n",
    "    def get_nuc_reads(self, number, nuc_pos, coverage_count=0): # checked\n",
    "        if coverage_count == 0:\n",
    "            x = (self.h5py_object[self.data[0,number-1]][nuc_pos-1, 1:self.get_coverage_count(number)+1])\n",
    "            return np.transpose(x)\n",
    "        else: \n",
    "            x = self.h5py_object[self.data[0,number-1]][nuc_pos-1, 1:(coverage_count+1)]\n",
    "            return np.transpose(x)\n",
    "    \n",
    "    def get_reads(self, number, coverage_count=0, true_len=0): # CHECK THE INDEXING\n",
    "        if coverage_count == 0 and true_len == 0:\n",
    "            _, true = self.get_barcode(number)\n",
    "            x = (self.h5py_object[self.data[0,number-1]][0:true, 1:self.get_coverage_count(number)+1])\n",
    "            return np.transpose(x)\n",
    "        else: \n",
    "            x = self.h5py_object[self.data[0,number-1]][0:true_len, 1:(coverage_count+1)]\n",
    "            return np.transpose(x)\n",
    "    \n",
    "    def get_qscores(self, number, coverage_count=0, true_len=0):\n",
    "        if coverage_count == 0 and true_len == 0:\n",
    "            _, true = self.get_barcode(number)\n",
    "            cov = self.get_coverage_count(number)\n",
    "            x = (self.h5py_object[self.data[0,number-1]][0:true, (1+cov):(1+2*cov)])\n",
    "            return np.transpose(x)\n",
    "        else: \n",
    "            x = self.h5py_object[self.data[0,number-1]][0:true_len, 1:(coverage_count+1)]\n",
    "            return np.transpose(x)        \n",
    "    \n",
    "    def get_interp_changes(self, number, coverage_count=0, true_len=0): # checked\n",
    "        if coverage_count == 0 and true_len == 0:\n",
    "            _, true = self.get_barcode(number)\n",
    "            x = self.h5py_object[self.data[0,number-1]][0:true,(1+2*self.get_coverage_count(number))]\n",
    "            return np.transpose(x)\n",
    "        else:\n",
    "            x = self.h5py_object[self.data[0,number-1]][0:true_len,(1+2*coverage_count)]\n",
    "            return np.transpose(x)\n",
    "    \n",
    "    def get_interp_mutations(self, number, coverage_count=0, true_len=0):\n",
    "        if coverage_count == 0 and true_len == 0:\n",
    "            _,true=self.get_barcode(number)\n",
    "            x = self.h5py_object[self.data[0,number-1]][0:true,(3+2*self.get_coverage_count(number))]\n",
    "            return np.transpose(x)\n",
    "        else:\n",
    "            x = self.h5py_object[self.data[0,number-1]][0:true_len,(3+2*coverage_count)]     \n",
    "            return np.transpose(x)      \n",
    "    \n",
    "    def get_interp_consensus(self, number, coverage_count=0, true_len=0):\n",
    "        if coverage_count == 0 and true_len == 0:\n",
    "            _, true= self.get_barcode(number)\n",
    "            x = self.h5py_object[self.data[0,number-1]][0:true,(2+2*self.get_coverage_count(number))]\n",
    "            return np.transpose(x) \n",
    "        else:\n",
    "            x = self.h5py_object[self.data[0,number-1]][0:true_len,(2+2*coverage_count)]\n",
    "            return np.transpose(x) \n",
    "    \n",
    "def seqblock_parser(seqblock):\n",
    "    seqblock_parsed = CleanSeqBlock()\n",
    "    total_columns = len(np.transpose(seqblock)) # is there a more efficient way?\n",
    "    total_rows = len(seqblock)\n",
    "    number_of_reads = int((total_rows - 4)/2) # last three rows have intreptations, first row is non-mutated target sequence, and N quality scores for N reads\n",
    "    \n",
    "    seqblock_parsed.cov = number_of_reads;\n",
    "    seqblock_parsed.len = total_columns;\n",
    "    \n",
    "    # print(total_columns)\n",
    "    # print(total_rows)\n",
    "    tag = \"\"\n",
    "    tag_length = 10_000\n",
    "    \n",
    "    for i in range(total_rows):\n",
    "        # print(f\" The sequenceblock input {seqblock}\") #debug\n",
    "        # print(f\" The m x n size of the seqblock {seqblock.shape}\") #debug\n",
    "        # print(f\" The ith row of the seqblock {current_seq}\") #debug\n",
    "        # print(f\" The m x n size of the sequence {current_seq.shape}\") #debug\n",
    "        \n",
    "        # new_format = np.chararray(total_columns) //CHANGED to string\n",
    "        stop_found = False\n",
    "        # new_format[:] = 'q' #debug\n",
    "        \n",
    "        # print(f\" The 0th entry of the initialized char array {new_format[0, 0]}\") #debug\n",
    "        # print(f\" Its row size {len(new_format[0])}\")\n",
    "        # if (i == 0):\n",
    "        #     original = \"\"\n",
    "        #     for j in range(total_columns):\n",
    "        #         original += chr(seqblock[i,j])\n",
    "                \n",
    "                \n",
    "        \n",
    "        # if (i != 0):\n",
    "  \n",
    "        if (i == 0 or i == 1):\n",
    "            new_format = \"\"\n",
    "            j=0\n",
    "        \n",
    "            while j < total_columns:\n",
    "                # print(j)\n",
    "                                        \n",
    "                if stop_found is False:\n",
    "                    if (seqblock[i,j] == ord('T') and (j+6) < total_columns):\n",
    "                        if (seqblock[i,j+1] == ord('A') and seqblock[i, j+2] == ord('A') and seqblock[i, j+3] == ord('T') and seqblock[i, j+4] == ord('A') and seqblock[i, j+5] == ord('G')):\n",
    "                            new_format += 'T'\n",
    "                            new_format += 'A'\n",
    "                            new_format += 'A'\n",
    "                            new_format += 'T'\n",
    "                            new_format += 'G'\n",
    "                            j = j+6\n",
    "                            stop_found = True\n",
    "                    if stop_found is False:\n",
    "                        new_format += chr(seqblock[i,j])\n",
    "                        j+=1\n",
    "            \n",
    "                elif stop_found is True and i == 1:\n",
    "                    tag += chr(seqblock[i,j])\n",
    "                    j+=1\n",
    "                else:\n",
    "                    j = total_columns\n",
    "            if (i == 1): \n",
    "                seqblock_parsed.barcode = tag\n",
    "                tag_length = len(tag)\n",
    "        else:\n",
    "            new_format = \"\"\n",
    "            for j in range(total_columns-tag_length):\n",
    "                new_format += chr(seqblock[i,j])\n",
    "                \n",
    "                \n",
    "        if (i == 0):\n",
    "            seqblock_parsed.target = new_format\n",
    "        elif (i == total_rows - 1): # last row is mutations 'x', total_rows is one more than total index\n",
    "            seqblock_parsed.interp_mutations = new_format\n",
    "        elif (i == total_rows - 2): # 2nd last is subjective consensus\n",
    "            seqblock_parsed.interp_consensus = new_format\n",
    "        elif (i == total_rows - 3): \n",
    "            seqblock_parsed.interp_changes = new_format\n",
    "        elif (i > 0 and i < number_of_reads + 1):\n",
    "            seqblock_parsed.reads.append(new_format)\n",
    "        else:\n",
    "            seqblock_parsed.qscores.append(new_format)\n",
    "    return seqblock_parsed     \n",
    "class CleanSeqBlock():\n",
    "    def __init__(self):\n",
    "        self.cov, self.len, self.target, self.interp_changes, self.interp_mutations, self.interp_consensus, self.barcode = 0,0,0,0,0,0,0\n",
    "        self.reads, self.qscores = [], []\n",
    "\n",
    "def pscore(Qscore): \n",
    "    return pow(10, -(Qscore-33)/10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# load_matlab_file function Testing\n",
    "file_name = 'consensus1.mat'\n",
    "variable_of_interest = 'align3'\n",
    "h5py_object, data = load_matlab_file(file_name, variable_name=variable_of_interest)\n",
    "\n",
    "# DNA_seqs can be used to grab each raw_seqblock\n",
    "DNA_seqs = DNA_SeqBlocks(h5py_object=h5py_object, data=data) \n",
    "\n",
    "FILE_SIZE = DNA_seqs.size #Constant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in barcodes and truelens for the day\n",
    "import pickle\n",
    "df = pickle.load(open('pickle/v1BarcodesTruelen.pickle', 'rb'))\n",
    "BARCODES = df.barcode\n",
    "TRUELENS = df.true_len"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feature: Using pandas dataframes and saving it in pickle. \n",
    "- The main dataset won't ever change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make sure it works on a small dataset first.\n",
    "tinyDS = range(1, 10001)\n",
    "DS = range(1, FILE_SIZE+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename each column for readability\n",
    "def rename(df, transpose=False):\n",
    "    if transpose == True: df = df.T\n",
    "    cov = int(0.5*(df.columns.size-4))\n",
    "    based_on_cov = ['target']\n",
    "    based_on_cov.extend([f'read{i}' for i in range(1, cov+1)])\n",
    "    based_on_cov.extend(x for x in [f'q{i}' for i in range(1, cov+1)])\n",
    "    based_on_cov.extend(['changes', 'contig', 'mutations'])\n",
    "    df.columns = based_on_cov\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Testing on a single seqblock\n",
    "import pandas as pd\n",
    "dflist = []\n",
    "\n",
    "# for num in tinyDS:\n",
    "for num in tinyDS:\n",
    "    true_len = TRUELENS[num]\n",
    "    df = pd.DataFrame(DNA_seqs.get_seqblock(num, true_len=true_len, df=True))\n",
    "    df = rename(df)\n",
    "    df.index.name = f'SB{num}' # Used to keep track of which specific sequence block the dataframe references\n",
    "    dflist.append(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(83, 1000, 0.083)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only return the sections with a z conflict\n",
    "chr = ord('z')\n",
    "\n",
    "\n",
    "conflicts = []\n",
    "for i in dflist:\n",
    "    test = i[i.changes == chr]\n",
    "    # if int(i.index.name[2:]) % 500 == 0: print(f'{test} \\n \\n')\n",
    "    if (test.size > 0): conflicts.append(test)\n",
    "print(f'\\n \\n')\n",
    "len(conflicts), len(dflist), len(conflicts)/len(dflist)\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      target  read1  read2  read3   q1  q2   q3  changes  contig  mutations\n",
      "SB13                                                                       \n",
      "506       67     67     65     67  126  75  126      122      67         32\n"
     ]
    }
   ],
   "source": [
    "print(conflicts[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Easy fix..."
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "7e45b5fe9540dcae422003c6a552f47067b28c6171ef36dc1ede64a9e99e7c78"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 ('venv')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
